### PR TITLE
Refactor klusterlet

### PR DESCRIPTION
**Pull request by qiujian16**
Signed-off-by: Jian Qiu <jqiu@redhat.com>

### CODE REVIEW

File: pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_pullsecret_reconcile.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	// Sync pull secret to the klusterlet addon namespace
+	// The reason we keep syncing secret instead of adding a label to trigger addonsecretcontroller to sync is:
+	// addonsecretcontroller only watch namespaces in the same cluster klusterlet is running on.
+	// And if addons are deployed in default mode on the managed cluster, but klusterlet is deployed in hosted on management cluster, then we still need to sync the secret here in klusterlet-controller using `managedClusterClients.kubeClient`.
+	err = r.syncPullSecret(ctx, r.kubeClient, r.managedClusterClients.kubeClient, klusterlet, addonNamespace, r.recorder)
+	if err != nil {
+		return klusterlet, reconcileStop, err
+	}
+
+	if config.InstallMode == operatorapiv1.InstallModeHosted {
```
Comment (zhujian7) [inline pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_pullsecret_reconcile.go]: Is it better to put this into the `managedReconcile`?

File: pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go [hunk 2]
```diff
@@ -117,6 +67,17 @@ type klusterletController struct {
...
 }
 
+type klusterletReconcile interface {
+	reconcile(ctx context.Context, cm *operatorapiv1.Klusterlet, config klusterletConfig) (*operatorapiv1.Klusterlet, reconcileState, error)
+}
+
+type reconcileState int64
+
+const (
+	reconcileStop reconcileState = iota
```
Comment (zhujian7) [inline pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:77]: Why do we need this? I observed that every use of reconcileStop is accompanied by an error

Comment (qiujian16) [reply pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go:77]: not necessarily, it is possible that we continue to run another reconcile with error. e.g. apply anyway but still return error so it check again later.

File: pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_pullsecret_reconcile.go (hunk 1 above)
Comment (qiujian16) [reply pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_pullsecret_reconcile.go]: yes, this reconcile is confusing, updated.

**Pull request merged**